### ISSUE TITLE
Additional Username/Password for connecting to VPN

### ISSUE THREAD

**Issue by ZephOne** [eduvpn/vpn-user-portal#182]
Is it possible to let the user generate a password protected private key to be used in the oepnvpn configuration ?

**Comment by fkooman** [comment]
No, that is not possible, not for the client, and neither for the server. 

Is there a particular reason why you'd need this?

**Comment by ZephOne** [comment]
Yep I wanted to protect the use of a configuration generated by a user via the user portal. If this configuration is stolen I don't want someone which is not the original user to use the vn.
But maybe it is something I have configure in openvpn configuration.  eg use LDAP or RADIUS authentification when a user wants to connect its openvpn client to openvpn server.

**Comment by fkooman** [comment]
Depending on your exact threat model (i.e. what/who do you want to protect against) there are various approaches...

**Comment by ZephOne** [comment]
I don't want an openvpn configuration to be usable by another user than the one who generated it.

**Comment by ZephOne** [comment]
Then I think that it should be configured on the vpn-server-node. Let's say I want to protect the use of an openvpn configuration via LDAP authentication, I think I should tell openvpn (maybe via vpn-server-node/config.php) to use the openvpn-auth-ldap plugin.

**Comment by fkooman** [comment]
Right, if you want to go that way that would be possible. This was part of eduVPN/Let's Connect! 1.0 actually but removed for 2.x as we want to move all authentication/authorization to the browser to support among other things different 2FA mechanisms and other VPN protocols.

A few servers I know about expire the OpenVPN certificate (and OAuth token) every 12 hours, so the user  is forced to login every day... maybe this could work as well, and is quite smooth when using the apps!

**Comment by ZephOne** [comment]
Thanks I'll see then. Just to write it here, if I use ldap authentication it is indeed not easy now with eduVPN/Let's Connect!:
- Manually add `auth-user-pass` to openvpn client config
- Manually add the use of the plugin openvpn-auth-ldap to the openvpn config if I use vpn-maint-apply-changes.

**Comment by fkooman** [comment]
So if I understand correctly you want to be able to have the user provide username/password when connecting to the VPN before being allowed to connect. This in addition to the client certificate they already have. 

This is not as easy as it seems, you also need to consider disconnects. The macOS and mobile apps disconnect all the time, and then you have to provide again the user/pass to connect.  You also need to consider `--auth-gen-token` or `--auth-gen-token-secret`.

**Comment by ZephOne** [comment]
> So if I understand correctly you want to be able to have the user provide username/password when connecting to the VPN before being allowed to connect. This in addition to the client certificate they already have.
Yes, exactly.


> This is not as easy as it seems, you also need to consider disconnects. The macOS and mobile apps disconnect all the time, and then you have to provide again the user/pass to connect.
Damn, I didn't know.
On macOS, is it only the case with the eduvpn client or does it also happen with blink ? Users using our current openvpn setup inside our school does not seem to complain about it but I can be wrong.

> You also need to consider --auth-gen-token or --auth-gen-token-secret
For OneTimePassword ?

**Comment by fkooman** [comment]
Right. If you just configure your VPN client to store the username/password then you are of course never asked for it again. But this assumes then that users are willing to share their VPN client configuration, but not their username/password, right?

Then it becomes a bit simpler indeed! Then what you proposed above will indeed work I think...

**Comment by ZephOne** [comment]
@fkooman, due to other things at work, I didn't take time to do any progress concerning this issue.
I take time yesterday to do it and made 3 pull requests which are mentioned in this issue. Let me know if you want me to open a new issue as this one is closed.

**Comment by fkooman** [comment]
Thanks! I'll get back to you soon!
